lib/codec.js

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+import {Base64PadCodec} from './codecs/base64Pad';
+import {CborldError} from './error';
+import {CodecMapCodec} from './codecs/codecMap';
+import {ContextCodec} from './codecs/context';
+import {SimpleTypeCodec} from './codecs/simpleType';
+import {UrlCodec} from './codecs/url';
+import {XsdDateTimeCodec} from './codecs/xsdDateTime';
+import {termCodecs} from './codecs/codecMap';
+
+/**
+ * Gets a list of all term codecs available to use when transforming term
+ * values.
+ *
+ * @returns {Array} - An array of all term codecs.
+ */
+export function getTermCodecs() {
+  return [...termCodecs.keys()].filter(k => typeof k === 'string');
+}
+
+/**
+ * Gets a term codec map for the given set of JSON-LD Context URLs. This map
+ * can be used to perform encoding or decoding.
+ *
+ * @param {object} options - The arguments to the function.
+ * @param {Array} options.contextUrls - The JSON-LD Context URLs to use when
+ *   building the term codec map.
+ * @param {object} options.documentLoader - The document loader to use when
+ *   dereferencing URLs.
+ * @param {object} [options.appTermMap] - A map of JSON-LD terms and
+ *   their associated CBOR-LD term codecs.
+ * @param {boolean} [options.encode] - If true, a map of term encoding codecs
+ *   are returned.
+ * @param {boolean} [options.decode] - If true, a map of term decoding codecs
+ *   are returned.
+ *
+ * @returns {object} - The term codec map for performing encoding or decoding.
+ */
+export async function getTermCodecMap({
+  contextUrls, appTermMap = new Map(), documentLoader, encode, decode
+} = {}) {
+  const termCodecMap = new Map();
+
+  // retrieve all JSON-LD Contexts
+  const contexts = await Promise.all(contextUrls.map(
+    async contextUrl => JSON.parse(await documentLoader(contextUrl))));
+
+  // generate a term codec map for each context
+  contexts.forEach(context => {
+    const map = _generateTermEncodingMap(context);
+    map.forEach((value, key) => termCodecMap.set(key, value));
+  });
+
+  // calculate the final values for the term codec map
+  // FIXME: ensure this is the same in all locales; may need to
+  // pick a particular locale to ensure it/do something else
+  const terms = Array.from(termCodecMap.keys()).sort();
+  terms.forEach((term, index) => {
+    if(encode) {
+      termCodecMap.get(term).value = index;
+    } else if(decode) {
+      termCodecMap.get(term).value = term;
+      termCodecMap.set(index, termCodecMap.get(term));
+    }
+  });
+
+  // overlay the app term map on the existing term encoding map
+  if(decode && appTermMap.size > 0) {
+    const codecMap = new CodecMapCodec();
+    codecMap.set({value: appTermMap, termCodecMap});
+    appTermMap = codecMap.decodeCBOR();
+  }
+
+  // clear terms used during encoding
+  if(decode) {
+    terms.forEach(term => termCodecMap.delete(term));
+  }
```

Comment:
I kept it separate because I was worried that people would miss line 78 and move it around. Yes, we should probably move it, but the entire function probably needs to be refactored. I added a FIXME in 754f22acd5047db709023049be5e2d44c0e11f26.